src/hypothesis/internal/validation.py

```diff
@@ -0,0 +1,146 @@
+    """Checks that value is either unspecified, or a valid integer.
+
+    Otherwise raises InvalidArgument.
+
+    """
+    if value is None:
+        return
+    check_type(integer_types, value)
+
+
+@check_function
+def check_valid_bound(value, name):
+    """Checks that value is either unspecified, or a valid interval bound.
+
+    Otherwise raises InvalidArgument.
+
+    """
+    if value is None or isinstance(value, integer_types + (Rational,)):
+        return
+    if math.isnan(value):
+        raise InvalidArgument(u'Invalid end point %s=%r' % (name, value))
+
+
+@check_function
+def try_convert(typ, value, name):
+    if value is None:
+        return None
+    if isinstance(value, typ):
+        return value
+    try:
+        return typ(value)
+    except TypeError:
+        raise InvalidArgument(
+            'Cannot convert %s=%r of type %s to type %s' % (
+                name, value, type(value).__name__, typ.__name__
+            )
+        )
+    except (OverflowError, ValueError):
+        raise InvalidArgument(
+            'Cannot convert %s=%r to type %s' % (
+                name, value, typ.__name__
+            )
+        )
+
+
+@check_function
+def check_valid_size(value, name):
+    """Checks that value is either unspecified, or a valid non-negative size
+    expressed as an integer/float.
+
+    Otherwise raises InvalidArgument.
+
+    """
+    if value is None:
+        return
+    check_type(integer_types + (float,), value)
+    if value < 0:
+        raise InvalidArgument(u'Invalid size %s=%r < 0' % (name, value))
+    if isinstance(value, float) and math.isnan(value):
+        raise InvalidArgument(u'Invalid size %s=%r' % (name, value))
+
+
+@check_function
+def check_valid_interval(lower_bound, upper_bound, lower_name, upper_name):
+    """Checks that lower_bound and upper_bound are either unspecified, or they
+    define a valid interval on the number line.
+
+    Otherwise raises InvalidArgument.
+
+    """
+    if lower_bound is None or upper_bound is None:
+        return
+    if upper_bound < lower_bound:
+        raise InvalidArgument(
+            'Cannot have %s=%r < %s=%r' % (
+                upper_name, upper_bound, lower_name, lower_bound
+            ))
+
+
+@check_function
+def check_valid_sizes(min_size, average_size, max_size):
+    check_valid_size(min_size, 'min_size')
+    check_valid_size(max_size, 'max_size')
+    check_valid_size(average_size, 'average_size')
+    check_valid_interval(min_size, max_size, 'min_size', 'max_size')
+    check_valid_interval(average_size, max_size, 'average_size', 'max_size')
+    check_valid_interval(min_size, average_size, 'min_size', 'average_size')
+
+    if average_size is not None:
+        if (
+            (max_size is None or max_size > 0) and
+            average_size is not None and average_size <= 0.0
+        ):
+            raise InvalidArgument(
+                'Cannot have average_size=%r with non-zero max_size=%r' % (
```

Comment:
I still don’t see what this is testing above the checks we already have. We know the sizes are all non-negative (by `check_valid_size`), and that min <= average <= max (by `check_valid_interval`).

If I’ve missed something and we do still need this, the conditionals can be simplified:

* On L141, we enter this block on L138 only if average_size is not None, so why recheck it again?
* Also L141, we know from earlier checks that average_size is non-negative. This condition could compare to 0.